kimchi/src/circuits/polynomials/foreign_field_add/witness.rs

```diff
@@ -231,32 +198,33 @@ fn init_foreign_field_add_rows<F: PrimeField>(
 ) {
     let left_row = 8 * index;
     let right_row = 8 * index + 4;
-    let witness_shape: [[WitnessCell<F>; COLUMNS]; 1] = [
+    let witness_shape: [[Box<dyn WitnessCell<F>>; COLUMNS]; 1] = [
         // ForeignFieldAdd row
         [
-            WitnessCell::Standard(CopyWitnessCell::create(left_row, 0)), // left_input_lo
-            WitnessCell::Standard(CopyWitnessCell::create(left_row + 1, 0)), // left_input_mi
-            WitnessCell::Standard(CopyWitnessCell::create(left_row + 2, 0)), // left_input_hi
-            WitnessCell::Standard(CopyWitnessCell::create(right_row, 0)), // right_input_lo
-            WitnessCell::Standard(CopyWitnessCell::create(right_row + 1, 0)), // right_input_mi
-            WitnessCell::Standard(CopyWitnessCell::create(right_row + 2, 0)), // right_input_hi
-            FieldElementCell::create(FieldElementType::Sign, 0),         // sign
-            FieldElementCell::create(FieldElementType::Overflow, 0),     // field_overflow
-            FieldElementCell::create(FieldElementType::Carry, LO),       // carry_lo
-            FieldElementCell::create(FieldElementType::Carry, MI),       // carry_mi
-            WitnessCell::Standard(ZeroWitnessCell::create()),
-            WitnessCell::Standard(ZeroWitnessCell::create()),
-            WitnessCell::Standard(ZeroWitnessCell::create()),
-            WitnessCell::Standard(ZeroWitnessCell::create()),
-            WitnessCell::Standard(ZeroWitnessCell::create()),
+            CopyCell::create(left_row, 0),      // left_input_lo
+            CopyCell::create(left_row + 1, 0),  // left_input_mi
+            CopyCell::create(left_row + 2, 0),  // left_input_hi
+            CopyCell::create(right_row, 0),     // right_input_lo
+            CopyCell::create(right_row + 1, 0), // right_input_mi
+            CopyCell::create(right_row + 2, 0), // right_input_hi
+            VariableCell::create("sign"),
+            VariableCell::create("overflow"), // field_overflow
+            VariableCell::create("carry_lo"),
+            VariableCell::create("carry_mi"),
+            ConstantCell::create(F::zero()),
+            ConstantCell::create(F::zero()),
+            ConstantCell::create(F::zero()),
+            ConstantCell::create(F::zero()),
+            ConstantCell::create(F::zero()),
         ],
     ];
 
-    for (row, wit) in witness_shape.iter().enumerate() {
-        for (col, cell) in wit.iter().enumerate() {
-            handle_ffadd_rows(witness, cell, (row, col), offset, sign, overflow, carry);
-        }
-    }
+    witness::init(
+        witness,
+        offset,
+        &witness_shape,
+        &variable_map!["sign" => sign, "overflow" => overflow, "carry_lo" => carry[0], "carry_mi" => carry[1]],
```

Comment:
Can't the variable map work over enums? or if you do then you have to explicitly address all of them (both the ones you use in this gate, and the ones you don't)?

Reply:
The old enums way was a lot more complex and needed a lot more code to do the same thing.  Also, I couldn't think of a way to make it generic, so you don't need to redefine types every time you want to do the same thing just for a new variable.